go/teams/transactions.go

```diff
@@ -397,28 +397,74 @@ func (tx *AddMemberTx) AddMemberByUsername(ctx context.Context, username string,
 	return err
 }
 
-// AddMemberByAssertion adds an assertion to the team.
+// preprocessAssertion takes an input assertion and determines if this is a valid Keybase-style assertion.
+// If it's an email (or phone) assertion, we assert that it only has one part (and isn't a+b compound).
+// If there is only one factor in the assertion, then that's returned. Otherwise, nil.
+func preprocessAssertion(m libkb.MetaContext, s string) (isEmail bool, single libkb.AssertionURL, err error) {
+	a, err := externals.AssertionParseAndOnly(m.G(), s)
+	if err != nil {
+		return false, nil, err
+	}
+	urls := a.CollectUrls(nil)
+	if len(urls) == 1 {
+		single = urls[0]
+	}
+	for _, u := range urls {
+		if u.IsServerTrust() {
+			isEmail = true
+		}
+	}
+	if isEmail && len(urls) > 1 {
+		return false, nil, NewMixedEmailAssertionError()
+	}
+	return isEmail, single, nil
+}
+
+// AddMemberByAssertionOrEmail adds an assertion to the team. It can handle
+// three major cases:
+//  1. joe OR joe+joe@reddit
+//  2. joe@reddit where
 // Does not attempt to resolve the assertion.
 // The return values (uv, username) can both be zero-valued if the assertion is not a keybase user.
-func (tx *AddMemberTx) AddMemberByAssertion(ctx context.Context, assertion string,
-	role keybase1.TeamRole) (username libkb.NormalizedUsername, uv keybase1.UserVersion, invite bool, err error) {
+func (tx *AddMemberTx) AddMemberByAssertionOrEmail(ctx context.Context, assertion string, role keybase1.TeamRole) (username libkb.NormalizedUsername, uv keybase1.UserVersion, invite bool, err error) {
 	team := tx.team
 	g := team.G()
 	m := libkb.NewMetaContext(ctx, g)
 
-	if kbun.NewNormalizedUsername(assertion).IsValid() {
-		upak, err := engine.ResolveAndCheck(m, assertion, true /* useTracking */)
+	defer m.CTrace(fmt.Sprintf("AddMemberTx.AddMemberByAssertionOrEmail(%s,%v) to team %q", assertion, role, team.Name()), func() error { return err })()
+
+	isEmail, single, err := preprocessAssertion(m, assertion)
+	if err != nil {
+		return "", uv, false, err
+	}
+
+	var doInvite bool
+	var upak keybase1.UserPlusKeysV2
+
+	if isEmail {
+		doInvite = true
```

Comment:
why are emails special? Couldn't they be resolve+checked if possible same as other assertions?

Reply:
maybe?  but i don't think they are